fix: add `ak.array_equal` to NEP18 overrides, documentation, and add one more test

tcawlfield opened the pull request:
* Now overrides numpy.array_equal
* Adding link to toctree
* Unit test checks numpy override

tcawlfield committed:
Fixing missed changes with ak.array_equal

* Now overrides numpy.array_equal
* Adding link to toctree
* Unit test checks numpy override

jpivarski reviewed (approved):
This is good; I agree that the heading `:caption:` should be "approximation and comparison".

The first set of test failures are due to the fact that Awkward only supports native-endian numbers, and you have some tests that make big-endian unsigned int32. I don't know what the second issue is.

Once those are fixed, this is mergeable!

tcawlfield committed:
np.array_equal no longer supports any numpy =? awkward

tcawlfield committed:
Fixing issue with ak_almost_equal and backend checks

Using ensure_same_backend now, which raises an exception instead of
returning False. This is a new behavior for ak.almost_equal.
This also removes a shortcut in the code, and we get a NotImplementedError
on a typetracer backend. Test_2678 updated to match this expectation.

jpivarski reviewed (approved):
This makes sense (the update in ec93fd2d9c76da5c5cb4cf6bfb9e5e82af3d784b): "same backend" is a broader criterion than someone is likely to be asking about when they ask whether two arrays are equal. Having it raise an error, like all of the other functions, is good for consistency.
